### ISSUE TITLE
rewrite dest port number

### ISSUE THREAD

**Issue by FOSquare** [xdp-project/xdp-tutorial#56]
Hello, I tried the rewriting dest port number. but can not load the bpf program.
the Errors are shown below.

sudo ./xdp_loader --dev veth-basic02 --force
libbpf: load bpf program failed: Invalid argument
libbpf: failed to load program 'xdp_port_rewrite'
libbpf: failed to load object 'xdp_prog_kern.o'
ERR: loading BPF-OBJ file(xdp_prog_kern.o) (-22): Invalid argument
ERR: loading file: xdp_prog_kern.o

The parse_udphdr() and xdp_port_rewrite_func() functions are shown. i can compile the C file without problem.
static __always_inline int parse_udphdr(struct hdr_cursor *nh,
                                        void *data_end,
                                        struct udphdr **udphdr)
{
        struct udphdr *udph = nh->pos;
        if (udph + 1 > data_end)
                return -1;

        nh->pos = udph + 1;
        *udphdr = udph;

        //return bpf_ntohs(udph->source);
        return bpf_ntohs(udph->dest);
}

SEC("xdp_port_rewrite")
int xdp_port_rewrite_func(struct xdp_md *ctx)
{
        void *data_end = (void *)(long)ctx->data_end;
        void *data = (void *)(long)ctx->data;
        struct ethhdr *eth;
        struct ipv6hdr *ipv6;
        struct udphdr *udp;

        struct hdr_cursor nh;
        int nh_type;
        int next_p;
        /* Start next header cursor position at data start */
        nh.pos = data;

        /* Packet parsing in steps: Get each header one at a time, aborting if
         * parsing fails. Each helper function does sanity checking (is the
         * header type in the packet correct?), and bounds checking.
         */
        nh_type = parse_ethhdr(&nh, data_end, &eth);

        /* Assignment additions go below here */

        next_p = parse_ip6hdr(&nh, data_end, &ipv6);

        //xdp_parser_func(struct xdp_md *ctx);
        int udp_source;
        udp_source = parse_udphdr(&nh, data_end, &udp);
       // udp =  nh.pos;
        udp->dest = bpf_htons(bpf_ntohs(udp->dest) - 1);
        return XDP_PASS;
}

Could you kindly help me to check it? Thanks so much

**Comment by aspsk** [comment]
Did you show the full source code? One problem with the code you posted is that you're not checking the return values of the parser functions.

**Comment by netoptimizer** [comment]
> sudo ./xdp_loader --dev veth-basic02 --force
> libbpf: load bpf program failed: Invalid argument
> libbpf: failed to load program 'xdp_port_rewrite'
> libbpf: failed to load object 'xdp_prog_kern.o'
> ERR: loading BPF-OBJ file(xdp_prog_kern.o) (-22): Invalid argument
> ERR: loading file: xdp_prog_kern.o

Did you get any "libbpf: -- BEGIN DUMP LOG ---" output?

`(-22): Invalid argument`

Is also given when verifier rejects your program... (thus this can easily be a programming error that the verifier found)

**Comment by FOSquare** [comment]
@aspsk 
Yes, I show all the source codes. Please kindly check the parse funciton. I did the checking as shown below. 

if (udph + 1 > data_end) 
       return -1;

For checking the return values of the parser functions, I do not quite understand, in the parse_icmp6hdr(),   parse_ip6hdr() and  parse_ethhdr() functions. I do not check the return values. And they work well.
 The functions are shown below.
****************************************
static __always_inline int parse_ethhdr(struct hdr_cursor *nh,
                                        void *data_end,
                                        struct ethhdr **ethhdr)
{
        struct ethhdr *eth = nh->pos;
        int hdrsize = sizeof(*eth);

        if (nh->pos + hdrsize > data_end)
                return -1;

        nh->pos += hdrsize;
        *ethhdr = eth;

        return bpf_ntohs(eth->h_proto);
}

/* Assignment 2: Implement and use this */

static __always_inline int parse_ip6hdr(struct hdr_cursor *nh,
                                        void *data_end,
                                        struct ipv6hdr **ip6hdr)
{
        struct ipv6hdr *ip6h = nh->pos;
        if (ip6h + 1 > data_end)
                return -1;

        nh->pos = ip6h + 1;
        *ip6hdr = ip6h;

        return bpf_ntohs(ip6h->nexthdr);
}

/* Assignment 3: Implement and use this */
static __always_inline int parse_icmp6hdr(struct hdr_cursor *nh,
                                          void *data_end,
                                          struct icmp6hdr **icmp6hdr)
{
        struct icmp6hdr *icmp6h = nh->pos;
        if (icmp6h + 1 > data_end)
                return -1;

        return  bpf_ntohs(icmp6h->icmp6_sequence);
}
****************************************

Thanks.

**Comment by FOSquare** [comment]
Thanks.
No, the above shows all the output. @netoptimizer

**Comment by aspsk** [comment]
> For checking the return values of the parser functions, I do not quite understand, in the parse_icmp6hdr(), parse_ip6hdr() and parse_ethhdr() functions. I do not check the return values. And they work well.

You are not checking what parsers return. How does your program handle an IPv4 TCP packet? The verifier will reject such code.

**Comment by FOSquare** [comment]
@aspsk
Thanks for your answer.
in the packet01 folder, Why no checking of tcp and udp works? Is it beacuse in tcp and udp cases, the icmp_index is 0???

SEC("xdp_packet_parser")
int  xdp_parser_func(struct xdp_md *ctx)
{
        void *data_end = (void *)(long)ctx->data_end;
        void *data = (void *)(long)ctx->data;
        struct ethhdr *eth;
        struct ipv6hdr *ipv6;
        struct icmp6hdr *icmp6hdr;
        int icmp_index;
        /* Default action XDP_PASS, imply everything we couldn't parse, or that
         * we don't want to deal with, we just pass up the stack and let the
         * kernel deal with it.
         */
        __u32 action = XDP_PASS; /* Default action */

        /* These keep track of the next header type and iterator pointer */
        struct hdr_cursor nh;
        int nh_type;
        int next_p;
        /* Start next header cursor position at data start */
        nh.pos = data;

        /* Packet parsing in steps: Get each header one at a time, aborting if
         * parsing fails. Each helper function does sanity checking (is the
         * header type in the packet correct?), and bounds checking.
         */
        nh_type = parse_ethhdr(&nh, data_end, &eth);
        if (nh_type != ETH_P_IPV6)
                goto out;

        /* Assignment additions go below here */

        next_p = parse_ip6hdr(&nh, data_end, &ipv6);

        icmp_index = parse_icmp6hdr(&nh, data_end, &icmp6hdr);

        if(icmp_index%2 == 0){
            action = XDP_PASS;
        }
        else{
            action = XDP_DROP;
        }
out:
        return xdp_stats_record_action(ctx, action); /* read via xdp_stats */
}

I tried to use the similar logic with udp_source, but it does not work. Can you give me a hint how can I check deal with the case of tcp? Do i must write one funciton? Thanks very much.
SEC("xdp_port_rewrite")
int xdp_port_rewrite_func(struct xdp_md *ctx)
{
        void *data_end = (void *)(long)ctx->data_end;
        void *data = (void *)(long)ctx->data;
        struct ethhdr *eth;
        struct ipv6hdr *ipv6;
        struct udphdr *udp;

        struct hdr_cursor nh;
        int nh_type;
        int next_p;
        /* Start next header cursor position at data start */
        nh.pos = data;

        /* Packet parsing in steps: Get each header one at a time, aborting if
         * parsing fails. Each helper function does sanity checking (is the
         * header type in the packet correct?), and bounds checking.
         */
        nh_type = parse_ethhdr(&nh, data_end, &eth);

        if (nh_type != ETH_P_IPV6)
                goto out;

        /* Assignment additions go below here */

        next_p = parse_ip6hdr(&nh, data_end, &ipv6);

        //xdp_parser_func(struct xdp_md *ctx);
        int udp_source;
        udp_source = parse_udphdr(&nh, data_end, &udp);

        if(udp_source ==0 )
                goto out;
       // udp =  nh.pos;
        udp->dest = bpf_htons(bpf_ntohs(udp->dest) - 1);
out:
        return XDP_PASS;
}

**Comment by aspsk** [comment]
Because this is a tutorial and the first assignment in the `packet01` is to add these checks https://github.com/xdp-project/xdp-tutorial/tree/master/packet01-parsing#assignment-1-fix-the-bounds-checking-error

**Comment by aspsk** [comment]
> @aspsk
> Thanks for your answer.
> in the packet01 folder, Why no checking of tcp and udp works? Is it beacuse in tcp and udp cases, the icmp_index is 0???

The original code in `packet01` doesn't parse tcp or udp. However, check how the return value from the `parse_ethhdr` is checked: https://github.com/xdp-project/xdp-tutorial/blob/master/packet01-parsing/xdp_prog_kern.c#L86 it assumes that code will work only with IPv6 packets. 

See an example of parsing UDP here: https://github.com/xdp-project/xdp-tutorial/blob/master/packet-solutions/xdp_prog_kern_02.c#L18
